### PR TITLE
LT-2509 Add contracts

### PR DESCRIPTION
**Pull request by IliyanIlievPH**

### CODE REVIEW

File: src/MarginTrading.AssetService.Contracts/MarketSettings/AddMarketSettingsRequest.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        /// <summary>
+        /// Id
+        /// </summary>
+        [Required]
+        public string Id { get; set; }
+        /// <summary>
+        /// MIC code
+        /// </summary>
+        [Required]
+        public string MICCode { get; set; }
```
Comment (gponomarev-lykke) [inline src/MarginTrading.AssetService.Contracts/MarketSettings/AddMarketSettingsRequest.cs:18]: Maybe we should get rid of abbreviations here

**Review by gponomarev-lykke — APPROVED** [review]

File: src/MarginTrading.AssetService.Contracts/IMarketSettingsApi.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        /// <summary>
+        /// Update market settings
+        /// </summary>
+        /// <returns></returns>
+        [Put("/api/market-settings/{id}")]
+        Task<ErrorCodeResponse<MarketSettingsErrorCodesContract>> UpdateMarketSettingsAsync(
+            [Body] UpdateMarketSettingsRequest request, [Required] string id);
+
+        /// <summary>
+        /// Add market settings
```
Comment (andrey-anokhin) [inline src/MarginTrading.AssetService.Contracts/IMarketSettingsApi.cs:46]: Copy-paste )

File: src/MarginTrading.AssetService.Contracts/IMarketSettingsApi.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        /// <summary>
+        /// Get market settings by id
+        /// </summary>
+        /// <returns></returns>
+        [Get("/api/market-settings/{id}")]
+        Task<GetMarketSettingsByIdResponse> GetByIdAsync(string id);
+
+        /// <summary>
+        /// Get all market settings ids
```
Comment (andrey-anokhin) [inline src/MarginTrading.AssetService.Contracts/IMarketSettingsApi.cs:23]: Here we return the full object, so the comment is missleading

**Review by andrey-anokhin — APPROVED** [review]

**Review by tarurar — APPROVED** [review]

File: src/MarginTrading.AssetService.SqlRepositories/Repositories/MarketSettingsRepository.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                    throw;
+                }
+            }
+        }
+
+        public async Task<Result<MarketSettingsErrorCodes>> UpdateAsync(MarketSettings model)
+        {
+            using (var context = _contextFactory.CreateDataContext())
+            {
+                var existing = await context.MarketSettings.FindAsync(model.Id);
```
Comment (andrey-anokhin) [inline src/MarginTrading.AssetService.SqlRepositories/Repositories/MarketSettingsRepository.cs:90]: We again getting the existing value 2 times. Let's try to remove that, please )

File: src/MarginTrading.AssetService.Contracts/MarketSettings/MarketSettingsChangedEvent.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        public string Username { get; set; }
+        [Key(1)]
+        public string CorrelationId { get; set; }
+        [Key(2)]
+        public string EventId { get; set; }
+        [Key(3)]
+        public DateTime Timestamp { get; set; }
+        [Key(4)]
+        [CanBeNull]
+        public MarketSettingsContract OldMarketSettings { get; set; }
```
Comment (tarurar) [inline src/MarginTrading.AssetService.Contracts/MarketSettings/MarketSettingsChangedEvent.cs:24]: Is it really needed to include the whole object? Or just made to be on the safe side for future features?

File: src/MarginTrading.AssetService.Services/MarketSettingsService.cs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            if (currentSettings == null)
+                return new Result<MarketSettingsErrorCodes>(MarketSettingsErrorCodes.MarketSettingsDoNotExist);
+
+            ValidateSettings(marketSettings);
+
+            //This is the current day taking into account the timezone
+            var currentDay = DateTime.UtcNow.Add(TZConvert.GetTimeZoneInfo(currentSettings.Timezone).BaseUtcOffset);
+
+            //Validate if we try to add holiday for already started trading day
+            if (marketSettings.Holidays.Select(x => x.Date.Date).Contains(currentDay.Date) &&
```
Comment (tarurar) [inline src/MarginTrading.AssetService.Services/MarketSettingsService.cs:82]: Shouldn't this code be inside `ValidateSettings` method?

File: src/MarginTrading.AssetService.Services/MarketSettingsService.cs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                currentSettings.Open <= currentDay.TimeOfDay &&
+                //Close will be Zero when it is set to 00h next day
+                (currentSettings.Close >= currentDay.TimeOfDay || model.Close == TimeSpan.Zero))
+            {
+                return new Result<MarketSettingsErrorCodes>(MarketSettingsErrorCodes.TradingDayAlreadyStarted);
+            }
+
+            await _marketSettingsRepository.UpdateAsync(marketSettings);
```
Comment (tarurar) [inline src/MarginTrading.AssetService.Services/MarketSettingsService.cs:90]: How do you handle conflicts like item has already been updated or even deleted after your last read operation?

File: src/MarginTrading.AssetService.Services/MarketSettingsService.cs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            if (model.DividendsShort < 0 || model.DividendsShort > 1)
+                return new Result<MarketSettingsErrorCodes>(MarketSettingsErrorCodes.InvalidDividendsShortValue);
+
+            if (model.Dividends871M < 0 || model.Dividends871M > 1)
+                return new Result<MarketSettingsErrorCodes>(MarketSettingsErrorCodes.InvalidDividends871MValue);
+
+            return new Result<MarketSettingsErrorCodes>();
+        }
+
+        private static MarketSettings GetMarketSettingsWithDefaults(MarketSettingsCreateOrUpdateDto model)
```
Comment (tarurar) [inline src/MarginTrading.AssetService.Services/MarketSettingsService.cs:138]: As far as I see this method doesn't have any reason to be a part of service. Looks like we better put it into domain model class.

File: src/MarginTrading.AssetService.SqlRepositories/Repositories/MarketSettingsRepository.cs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+                context.MarketSettings.Update(existing);
+
+                try
+                {
+                    await context.SaveChangesAsync();
+                    return new Result<MarketSettingsErrorCodes>();
+                }
+                catch (DbUpdateConcurrencyException e)
+                {
+                    if (e.Message.Contains(DoesNotExistException))
```
Comment (tarurar) [inline src/MarginTrading.AssetService.SqlRepositories/Repositories/MarketSettingsRepository.cs:105]: Do we have other than "text comparison" options to check for this kind of error? I mean, like sql error code or something?

Comment (IliyanIlievPH) [reply src/MarginTrading.AssetService.SqlRepositories/Repositories/MarketSettingsRepository.cs:105]: Actually, it cannot be cast to SQL exception and does not have sql code

File: src/MarginTrading.AssetService.Services/MarketSettingsService.cs (hunk 8 above)
Comment (IliyanIlievPH) [reply src/MarginTrading.AssetService.Services/MarketSettingsService.cs:138]: agree

File: src/MarginTrading.AssetService.Contracts/MarketSettings/MarketSettingsChangedEvent.cs (hunk 5 above)
Comment (IliyanIlievPH) [reply src/MarginTrading.AssetService.Contracts/MarketSettings/MarketSettingsChangedEvent.cs:24]: for future features

**Pull request merged**